File: swift/question.swift
import Swift
import Foundation

func question(prompt: String, valid: [String]) -> String {
	let joinedValid = valid.joined(separator: ", ")

	while true {
		print(prompt)
		if !valid.isEmpty {
			print("(\(joinedValid))", terminator: "")
		}
		print(": ", terminator: "")

		if let inp = readLine() {
			if valid.isEmpty {
				return inp
			}

			for s in valid {
				if s == inp {
					return s
				}
			}

			print("\"\(inp)\" is not a valid answer!")
		} else {
			print("Error reading from stdin")
			abort()
		}
	}
}

print(question(prompt: "Foo", valid: ["Bar", "Baz"]))

File: julia/question.jl
# implementation of the question function in julia
function question(prompt::AbstractString, valid::AbstractVector{<:AbstractString})
    loopprompt = isempty(valid) ? "" : '(' * join(valid, ", ") * ')'
    while true
        print(prompt, '\n', loopprompt, ": ")
        inp = readline(stdin)
        (isempty(valid) || inp in valid) && return inp
        println("\"$inp\" is not a valid answer", '\n')
    end
end

question("foo", ["bar", "baz"])
